### ISSUE TITLE
Comm panel template UI

### ISSUE THREAD

**Issue by willgearty** [learning-unlimited/ESP-Website#2885]
Admins at SplashCon 2020 suggested adding functionality to the comm panel that allows you to pick from a selection of templates (like in other services like MailChimp) that would then populate the email field with the HTML code for that template.

Here's a mock-up of what it might look like:
![mockup](https://user-images.githubusercontent.com/7232514/74210403-3ac0d780-4c51-11ea-8f91-b0db944fd6fd.png)

**Comment by willgearty** [comment]
Copying this from https://github.com/learning-unlimited/ESP-Website/pull/1898#issuecomment-584470422 since that PR will probably not include a UI:

> We could have some array of buttons like they do in other mass-mailing services (e.g. Mailchimp) that would import/inject a template into the commpanel GUI. The most basic version could pull all of templates from a template folder (e.g. /program/commpanel/templates) and then just show the name of the template in the button. A more advanced version could actually render the templates within the buttons. This way, chapters could create any number of custom templates (as template overrides) within this folder, and each one would be selectable from the comm panel.

**Comment by milescalabresi** [comment]
I think a way to get here is using third-party integration. Some reasonable-looking options include:

1. https://www.usewaypoint.com/open-source/emailbuilderjs
2. https://www.limejourney.com/
3. https://grapesjs.com/ (also has a web-page builder, which could be interesting for custom QSDs/template overrides/themes/something)
4. https://topol.io/ (closed source)

**Comment by willgearty** [comment]
emailbuilderjs looks really good!

**Comment by milescalabresi** [comment]
Yeah! My only question is whether the grapejs page builder is nice enough that we'd want to have that in our back pocket for a future enhanced theme/QSD editor. If you think that's sufficiently far off, I'm happy to do emailbuilderjs

**Comment by willgearty** [comment]
As far as I can tell, grapejs would be a lot of work to implement. Definitely not worth the investment at the moment, but feel free to make a separate issue about implementing a template override/QSD builder

**Comment by Eshaan-byte** [comment]
Hi @willgearty, can I work on this issue? Thank you!

**Comment by willgearty** [comment]
@Eshaan-byte Thanks for your interest in contributing to our project! I've assigned you to this issue. Please make sure to check out our [instructions for getting started](https://github.com/learning-unlimited/ESP-Website/wiki#getting-started) and follow our [contribution rules](https://github.com/learning-unlimited/ESP-Website/wiki#rules). Let me know if you have any questions.

**Comment by dev-gandhi** [comment]
Hey @willgearty, I noticed the task was assigned to @Eshaan-byte. Does that mean it is not available for me?

I had love to contribute to this issue!

**Comment by willgearty** [comment]
That is correct. This task is already assigned. We appreciate your interest in contributing to our project, but please leave this task for the assignee to work on.

**Comment by Eshaan-byte** [comment]
Hi, @willgearty and @milescalabresi , a couple of questions before I start:

-Should this PR implement a simple template selector (e.g., a dropdown/buttons to pick from existing templates like default and minimal), or should it integrate emailbuilderjs as a drag-and-drop builder?

-Does Step 2 of #3803 (allowing chapters to supply multiple templates) need to be completed first, or should I build the UI to work with whatever templates currently exist?

**Comment by dev-gandhi** [comment]
> That is correct. This task is already assigned. We appreciate your interest in contributing to our project, but please leave this task for the assignee to work on.

Thanks for the confirmation @willgearty!, is there any other task for me? (Good first issue)

**Comment by willgearty** [comment]
@Eshaan-byte I think the minimum viable product for this issue is a simple template selector. The stretch goal (and eventual goal) is a drag-and-drop builder like emailbuilderjs. For the simple template selector, I think we would provide a set of templates that could be used, and then chapters would be able to supply their own somehow down the line. For the drag-and-drop builder, the chapter-specific templates would be added separately afterwards.

@dev-gandhi you can check out our current [milestone](https://github.com/learning-unlimited/ESP-Website/milestone/25) or filter the issues to those with ["Starter", "Small", and/or "Easy" tag(s)](https://github.com/learning-unlimited/ESP-Website/issues?q=is%3Aissue%20state%3Aopen%20label%3AStarter%20OR%20label%3AEasy%20OR%20label%3ASmall).

**Comment by Eshaan-byte** [comment]
Hi @willgearty, I've been working on this. I have two versions ready:

Version A — Simple: Clean cards showing just the template name (Default, Minimal)

Version B — Detailed: Cards with descriptions and mini CSS-based layout previews showing the template structure (e.g., Default shows a footer section, Minimal doesn't)

<img width="1440" height="859" alt="Image" src="https://github.com/user-attachments/assets/2197941c-6e90-42f1-90d6-af04feeda774" />

Both versions auto-discover *_email.html files from templates/email/, and selection persists through compose → preview → back.

Some  questions:
-Which version do you prefer, or something in between?
-Should I add more built-in templates beyond Default and Minimal, or is the selector UI with these two sufficient for the MVP?

**Comment by willgearty** [comment]
This looks great! I think the version with the descriptions and mini-CSS is better. Some thoughts:
- all templates should have the unsubscribe footer
- does clicking on the template load it into the text box below? Or set it as the template for when the text is processed? If it's the latter, I'm not sure how many templates would make sense. What can we reasonably vary at the moment if all the user can edit is the main text?

**Comment by Eshaan-byte** [comment]
Thanks for the feedback! To clarify — clicking a template card sets it as the wrapper used when the email is processed. The user's body text gets inserted into the template's {{msgbody}} slot, so it doesn't load anything into the text editor.

Good point about the unsubscribe footer — I'll make sure all templates include it. Given that, would you prefer:

A) Templates that vary the visual wrapper around the body (e.g., branded header, color scheme)
B) Templates that load pre-built HTML into the text editor as a starting point
C) Something else?

**Comment by willgearty** [comment]
Unassigning @Eshaan-byte because this issue is part of a GSoC project. Please submit a GSoC project proposal if you would like to work on this issue.

**Comment by willgearty** [comment]
@Eshaan-byte I think A is the way to go for a MVP.

**Comment by Eshaan-byte** [comment]
Thanks @willgearty, understood! I'll work on my GSoC proposal for this. In the meantime, I'll pick up a separate issue for my code contribution.

**Comment by nithin434** [comment]
I have a plan to work on this issue as I previously developed a similar application for the agentic system to send emails.

I’m particularly interested in extending the template system to support editing existing templates, adding images and richer layout controls (like configurable message-body sections), and exploring an agent-assisted workflow that can generate custom email templates from a simple idea or prompt. This would help admins quickly create branded, reusable templates while keeping the current MVP approach intact. 

Please let me know if this direction aligns with the project goals. I would like to work on this for my gsoc @willgearty

**Comment by willgearty** [comment]
@nithin434 I don't think we have the infrastructure and/or budget to directly support any sort of agent-assisted tools within the site. However, you could let chapters use their own API key as a `Tag` if they'd like to use an agent within the email builder. Otherwise, the rest of your plan certainly aligns with the project goals. Happy to answer any other questions you have and looking forward to reviewing your proposal.

**Comment by nithin434** [comment]
Thanks for the clarification, that approach works well. I’ll focus first on the admin template builder (editing existing templates, image support, configurable body sections), and keep the agent feature optional through a user-provided API key ( OpenRouter with selectable base model).
I’ll begin working on this and share draft updates for feedback. Please let me know if I can proceed with this as my GSoC project direction. @willgearty

**Comment by willgearty** [comment]
@nithin434 Please note that this issue is part of a GSoC project. Please submit a GSoC project proposal if you would like to work on this issue. What you've described would be a good approach for a proposal.

**Comment by nithin434** [comment]
Thanks for the note! I’ll submit this as a GSoC proposal and refine the implementation details accordingly.

**Comment by Shofikul-Isl4m** [comment]
Hey @willgearty, do you have a Slack channel I could join?

**Comment by SuyashJain17** [comment]
@Shofikul-Isl4m they are working on it

**Comment by willgearty** [comment]
@Shofikul-Isl4m @SuyashJain17 see https://github.com/learning-unlimited/ESP-Website/discussions/4310

**Comment by MYMISERABLEMIND17** [comment]
Hi @willgearty! I'm interested in working on the Comm panel template UI as part of GSoC 2026.

After reading through the issue thread and the existing comm panel code, I think the **emailbuilderjs** integration is the right direction for MVP. My proposed approach:

**Phase 1 — MVP Template Selector:**
- Add a template picker UI to the comm panel that reads from `/program/commpanel/templates/`
- Display template names as selectable buttons (as described in the issue)
- On selection, inject the template HTML into the existing comm panel editor
- Allow chapters to add custom templates via template overrides

**Phase 2 — emailbuilderjs Integration:**
- Embed emailbuilderjs as a visual drag-and-drop email builder
- Support adding images, configuring layout sections, and editing template content
- Save/export the built email as HTML back into the comm panel

**Phase 3 — Polish:**
- Preview rendering of templates within the selector buttons
- Ensure all templates are mobile-responsive
- Write tests for the new template loading logic

I'm currently setting up the dev environment and working on smaller issues to get familiar with the codebase. I will submit a formal GSoC 2026 proposal with a detailed week-by-week timeline.

Happy to discuss further — thank you!

**Comment by omku415** [comment]
Hey,@willgearty
I’ve reviewed the **Communications Improvements** project for the ESP Website. My understanding is that the current communication panel allows administrators to send emails, but it lacks modern features such as reusable templates, scheduled sending, and teacher-level communication tools.

My approach would be to extend the Django-based system by introducing a template management module where admins can create, edit, and preview reusable HTML email templates stored in the database. For scheduled sending, I would implement a scheduling mechanism using SendGrid’s scheduling API or a cron/background task system to queue emails for delivery at a specified date and time.

Additionally, I would develop a teacher-facing communication panel that allows teachers to easily email students enrolled in their classes by leveraging existing enrollment data. On the frontend, I would design UI components for template previews, a date/time picker for scheduling emails, and a dashboard to view and manage pending or scheduled emails.

These improvements align with issues #2885 (Comm panel template UI), #3833 (Teacher communication panel), and #3951 (Scheduled email sending).

**Comment by workonlly** [comment]
\assign

**Comment by ishandabi854-pixel** [comment]
# PR Draft for Issue #2885: Comm panel template UI

## Overview
This PR implements a database-driven email template system for the Communications Panel. It allows administrators to create, edit, and preview email templates through the database instead of relying on hardcoded file templates.

## Related Issues
- #2885: Comm panel template UI
- #3833: Teacher communication panel
- #3951: Scheduled email sending (prerequisite infrastructure)

## Why this method is used?
Currently, the ESP Website uses static HTML files in the `templates/email/` directory for communication panel emails. This approach has several drawbacks:
1. **Lack of Flexibility**: Admins cannot create or modify templates without developer assistance.
2. **Maintenance Overhead**: Adding a new template requires a code change and deployment.
3. **No Preview**: There is no easy way to preview how a template will look with dynamic data before it is sent.

By moving templates to the database:
- **Admin Empowerment**: Administrators can manage communication templates through the standard Django admin interface.
- **Real-time Updates**: Changes to templates are immediate and don't require server restarts.
- **Prerequisite for Scheduling**: Database-driven templates are necessary for scheduling emails, as the template content needs to be persisted alongside the scheduled task.

## Implementation Details

### 1. Database Model (`esp/esp/dbmail/models.py`)
Introduced a new `EmailTemplate` model that stores:
- `name`: Unique identifier for the template.
- `subject`: Default subject line.
- `body_html`: The HTML content of the template, supporting Django template syntax and the `{{ msgbody }}` placeholder.

### 2. Admin Interface (`esp/esp/dbmail/admin.py`)
Registered the `EmailTemplate` model so it is accessible to site administrators.

### 3. Communications Panel Logic (`esp/esp/program/modules/handlers/commmodule.py`)
- Updated the template selection logic to prioritize database-driven templates.
- Added a fallback to existing file-based templates for backward compatibility.
- Enhanced the preview functionality to render database templates correctly with dynamic user and program data.

---

## Files Changed

### `esp/esp/dbmail/models.py`
Add the `EmailTemplate` model:
```python
class EmailTemplate(models.Model):
    name = models.CharField(max_length=255, unique=True)
    description = models.TextField(blank=True, null=True)
    subject = models.CharField(max_length=255, help_text="Default subject for emails using this template")
    body_html = models.TextField(help_text="The HTML content of the template. Use {{ msgbody }} for the message content.")
    
    def __str__(self):
        return self.name
```

### `esp/esp/dbmail/admin.py`
Register the new model:
```python
from esp.dbmail.models import EmailTemplate

class EmailTemplateAdmin(admin.ModelAdmin):
    list_display = ('name', 'subject')
    search_fields = ('name', 'subject', 'body_html')
admin_site.register(EmailTemplate, EmailTemplateAdmin)
```

### `esp/esp/program/modules/handlers/commmodule.py`
Modify `commprev` and `commfinal` to handle `EmailTemplate` objects:
```python
# In commprev and commfinal:
from esp.dbmail.models import EmailTemplate

template_id = request.POST.get('template')
try:
    db_template = EmailTemplate.objects.get(id=template_id)
    rendered_text = Template(db_template.body_html).render(DjangoContext({'msgbody': body}))
except (EmailTemplate.DoesNotExist, ValueError):
    # Fallback to file-based templates
    rendered_text = render_to_string('email/{}_email.html'.format(template_id), {'msgbody': body})
```

**Comment by willgearty** [comment]
@ishandabi854-pixel This issue is part of a GSoC project. Please submit a GSoC project proposal if you would like to work on this issue.

**Comment by Anchal1808** [comment]
Hi, I’m Anchal, a CSE student preparing for GSoC 2026.
I’m interested in working on this issue (#2885 - Comm panel template UI).
From my understanding, we need to add template selection buttons in the comm panel that can inject predefined templates into the email editor.
I’d like to start with a basic version by implementing simple template selection buttons and template injection functionality. Does this sound like a good starting point?
Thanks!

**Comment by nithin434** [comment]
@Anchal1808 This is supposed to be done in gsoc. You need to submit gsoc proposal around your implementation plan to work on it.

**Comment by Anchal1808** [comment]
Thanks for the clarification!
I understand that this is intended as a GSoC project. I’m really interested in working on it and will be preparing a proposal for it.
In the meantime, I’d like to start contributing and get familiar with the codebase. Would it be okay if I begin with a smaller part related to this (like a basic UI prototype), or could you suggest some beginner-friendly issues to start with?
Thanks!

**Comment by nithin434** [comment]
You can look for opened issues to work on. Depends on your skills, you can read different issues description to find your domain relevant beginner friendly issue and try to simulate it on your local system and before starting to work try to understand issue correctly.
